src/ctap/apdu.rs

```diff
@@ -119,48 +123,103 @@ impl TryFrom<&[u8]> for APDU {
+                return Ok(APDU {
+                    header: array_ref!(header, 0, APDU_HEADER_LEN).into(),
+                    lc: byte_0.into(),
+                    data: payload[1..(payload.len() - 1)].to_vec(),
+                    le: if last_byte == 0x00 { 0x100 } else { last_byte },
+                    case_type: ApduType::Short(Case::Lc1DataLe1),
+                });
             }
-            if payload.len() == (1 + byte_0 + 1) as usize && byte_0 != 0 {
-                apdu.case_type = ApduType::Short(Case::LcDataLe);
-                apdu.lc = byte_0.into();
-                apdu.data = payload[1..(payload.len() - 1)].to_vec();
-                apdu.le = (*payload.last().unwrap()).into();
-                if apdu.le == 0x00 {
-                    apdu.le = 0x100;
+            if payload.len() > 2 {
+                // Lc is possibly three-bytes long
+                let extended_apdu_lc: usize = BigEndian::read_u16(&payload[1..]) as usize;
+                let extended_apdu_le_len: usize = if payload.len() > extended_apdu_lc {
+                    payload.len() - extended_apdu_lc - 3
+                } else {
+                    0
+                };
+                if byte_0 == 0 && extended_apdu_le_len <= 3 {
+                    // If first byte is zero AND the next two bytes can be parsed as a big-endian
+                    // length that covers the rest of the block (plus few additional bytes for Le), we
+                    // have an extended-length APDU
+                    let last_byte: u32 = (*payload.last().unwrap()).into();
+                    return Ok(APDU {
+                        header: array_ref!(header, 0, APDU_HEADER_LEN).into(),
+                        lc: extended_apdu_lc as u16,
+                        data: payload[3..(payload.len() - extended_apdu_le_len)].to_vec(),
+                        le: match extended_apdu_le_len {
+                            0 => 0,
+                            1 => {
+                                if last_byte == 0x00 {
+                                    0x100
+                                } else {
+                                    last_byte
+                                }
+                            }
+                            2 => BigEndian::read_u16(
+                                &payload[payload.len() - extended_apdu_le_len..],
+                            ) as u32,
+                            3 => BigEndian::read_u32(
+                                &payload[payload.len() - extended_apdu_le_len..],
+                            ),
+                            _ => 0,
+                        },
+                        case_type: ApduType::Extended(match extended_apdu_le_len {
+                            0 => Case::Lc3Data,
+                            1 => Case::Lc3DataLe1,
+                            2 => Case::Lc3DataLe2,
+                            3 => Case::Le3,
+                            _ => Case::Unknown,
```

Comment:
While I was thinking about `Unknown`, I was wondering why we return an unknown case, and not
`Err(ApduStatusCode::SW_COND_USE_NOT_SATISFIED);`
?

Reply:
Maybe I just don't see the use case? It makes the logic uses your APDUs a bit simpler if you remove the enum variant. If we don't need it?

Reply:
Addressed in 71ec2cf937260c075222960405f478108cbe0106